<h1>ACTIVITY1 - PALMER PENGUINS DATA ANALYTICS PROJECT</h1>

<h2>Analyst: Gerald Prio</h2>

In [1]:
#libraries

In [2]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [3]:
penguin  = pd.read_csv('datasets\\penguins.csv')

In [4]:
penguin

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

<h1>Descriptive Analytics</h1>

In [5]:
penguin.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

<h1>Filtering Column/Data</h1>

In [6]:
penguin['species']

0         Adelie
1         Adelie
2         Adelie
3         Adelie
4         Adelie
         ...    
339    Chinstrap
340    Chinstrap
341    Chinstrap
342    Chinstrap
343    Chinstrap
Name: species, Length: 344, dtype: object

In [7]:
adelie_df = penguin[penguin['species'] == "Adelie"] 

In [8]:
adelie_df

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
149  Adelie      Dream            37.8           18.1              193.0   
150  Adelie      Dream            36.0           17.1              187.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
147       3475.0  female  2009  
148       3450.0  female  2009  
149       3750.0    male  2009  
150       3700.0  female  2009  
151       4000.0    male  2009  

[152 rows x 8 columns]

In [9]:
penguin['body_mass_g'] >=4000

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340    False
341    False
342     True
343    False
Name: body_mass_g, Length: 344, dtype: bool

In [10]:
(penguin['body_mass_g'] >=4000) & (penguin['species'] == 'Adelie')

0      False
1      False
2      False
3      False
4      False
       ...  
339    False
340    False
341    False
342    False
343    False
Length: 344, dtype: bool

In [11]:
penguin['bill_length_mm'] <=40

0       True
1       True
2      False
3      False
4       True
       ...  
339    False
340    False
341    False
342    False
343    False
Name: bill_length_mm, Length: 344, dtype: bool

In [12]:
penguin['flipper_length_mm'] >= 200

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340     True
341    False
342     True
343    False
Name: flipper_length_mm, Length: 344, dtype: bool

<h1>Cleaning the Data Frame</h1>

In [13]:
penguin.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h1>What to do with all NULL VALUES?</h1>

In [14]:
penguin['bill_length_mm'].mean()

np.float64(43.9219298245614)

In [15]:
penguin['sex'].mode()

0    male
Name: sex, dtype: object

<h1>How to FILLUP NULL VALUES?</h1>

In [16]:
penguin['bill_length_mm'] = penguin['bill_length_mm'].fillna(penguin['bill_length_mm'].mean())

In [17]:
penguin['sex'] = penguin['sex'].fillna(penguin['sex'].mode()[0])

In [18]:
penguin.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        2
flipper_length_mm    2
body_mass_g          2
sex                  0
year                 0
dtype: int64

In [19]:
penguin['flipper_length_mm'] = penguin['flipper_length_mm'].fillna(penguin['flipper_length_mm'].mean())

In [20]:
penguin.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        2
flipper_length_mm    0
body_mass_g          2
sex                  0
year                 0
dtype: int64

<h1>DATA TYPES</h1>

In [21]:
penguin.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      344.000000     342.000000         344.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.443643       1.974793          14.020657   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.275000      15.600000         190.000000  3550.000000   
50%         44.250000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

<h3>Changing Data Types</h3>

In [22]:
penguin['species'] = penguin['species'].astype('category')
penguin['sex'] = penguin['sex'].astype('category')

In [23]:
penguin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    object  
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      342 non-null    float64 
 4   flipper_length_mm  344 non-null    float64 
 5   body_mass_g        342 non-null    float64 
 6   sex                344 non-null    category
 7   year               344 non-null    int64   
dtypes: category(2), float64(4), int64(1), object(1)
memory usage: 17.2+ KB


In [24]:
penguin.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      344.000000     342.000000         344.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.443643       1.974793          14.020657   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.275000      15.600000         190.000000  3550.000000   
50%         44.250000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

<h3>Which penguins have flipper length between 200 mm and 210 mm?</h3>

In [25]:
penguin[(penguin['flipper_length_mm'] >=200) & (penguin['flipper_length_mm'] <=210)]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
3       Adelie  Torgersen        43.92193            NaN         200.915205   
53      Adelie     Biscoe        42.00000           19.5         200.000000   
90      Adelie      Dream        35.70000           18.0         202.000000   
91      Adelie      Dream        41.10000           18.1         205.000000   
95      Adelie      Dream        40.80000           18.9         208.000000   
101     Adelie     Biscoe        41.00000           20.0         203.000000   
123     Adelie  Torgersen        41.40000           18.5         202.000000   
129     Adelie  Torgersen        44.10000           18.0         210.000000   
137     Adelie      Dream        40.20000           20.1         200.000000   
151     Adelie      Dream        41.50000           18.5         201.000000   
154     Gentoo     Biscoe        48.70000           14.1         210.000000   
157     Gentoo     Biscoe        46.50000           13.5         210.000000   
160     Gentoo     Biscoe        43.30000           13.4         209.000000   
166     Gentoo     Biscoe        45.80000           14.6         210.000000   
168     Gentoo     Biscoe        42.00000           13.5         210.000000   
170     Gentoo     Biscoe        46.20000           14.5         209.000000   
180     Gentoo     Biscoe        48.20000           14.3         210.000000   
183     Gentoo     Biscoe        42.80000           14.2         209.000000   
184     Gentoo     Biscoe        45.10000           14.5         207.000000   
190     Gentoo     Biscoe        44.00000           13.6         208.000000   
191     Gentoo     Biscoe        48.70000           15.7         208.000000   
192     Gentoo     Biscoe        42.70000           13.7         208.000000   
194     Gentoo     Biscoe        45.30000           13.7         210.000000   
198     Gentoo     Biscoe        45.50000           13.9         210.000000   
202     Gentoo     Biscoe        46.60000           14.2         210.000000   
204     Gentoo     Biscoe        45.10000           14.4         210.000000   
208     Gentoo     Biscoe        43.80000           13.9         208.000000   
210     Gentoo     Biscoe        43.20000           14.5         208.000000   
212     Gentoo     Biscoe        45.30000           13.8         208.000000   
228     Gentoo     Biscoe        47.50000           14.2         209.000000   
250     Gentoo     Biscoe        48.40000           14.4         203.000000   
258     Gentoo     Biscoe        41.70000           14.7         210.000000   
260     Gentoo     Biscoe        43.30000           14.0         208.000000   
261     Gentoo     Biscoe        48.10000           15.1         209.000000   
271     Gentoo     Biscoe        43.92193            NaN         200.915205   
289  Chinstrap      Dream        52.00000           18.1         201.000000   
291  Chinstrap      Dream        50.50000           19.6         201.000000   
302  Chinstrap      Dream        50.50000           18.4         200.000000   
303  Chinstrap      Dream        49.50000           19.0         200.000000   
305  Chinstrap      Dream        52.80000           20.0         205.000000   
307  Chinstrap      Dream        54.20000           20.8         201.000000   
309  Chinstrap      Dream        51.00000           18.8         203.000000   
313  Chinstrap      Dream        52.00000           20.7         210.000000   
315  Chinstrap      Dream        53.50000           19.9         205.000000   
316  Chinstrap      Dream        49.00000           19.5         210.000000   
321  Chinstrap      Dream        50.80000           18.5         201.000000   
327  Chinstrap      Dream        51.40000           19.0         201.000000   
329  Chinstrap      Dream        50.70000           19.7         203.000000   
333  Chinstrap      Dream        49.30000           19.9         203.000000   
334  Chinstrap      Dream        50.20000           18.8 

<h3>Which year had the heaviest penguin recorded, and what species was it?</h3>

<h1>Data Aggregation</h1>

In [26]:
#species,,bill_length_mm, bill_depth_mm, flipper_length_ mm

sbbf = penguin.groupby(['species', 'island', 'bill_length_mm','bill_depth_mm']).size()

In [27]:
sbbf

species  island     bill_length_mm  bill_depth_mm
Adelie   Biscoe     32.1            13.1             0
                                    13.2             0
                                    13.3             0
                                    13.4             0
                                    13.5             0
                                                    ..
Gentoo   Torgersen  59.6            20.7             0
                                    20.8             0
                                    21.1             0
                                    21.2             0
                                    21.5             0
Length: 118800, dtype: int64

<h3>What is the average flipper length of every species in all island?</h3>

In [28]:
avg_flip = penguin.groupby(['species', 'island'])['flipper_length_mm'].mean()

In [29]:
avg_flip

species    island   
Adelie     Biscoe       188.795455
           Dream        189.732143
           Torgersen    191.382985
Chinstrap  Biscoe              NaN
           Dream        195.823529
           Torgersen           NaN
Gentoo     Biscoe       217.055768
           Dream               NaN
           Torgersen           NaN
Name: flipper_length_mm, dtype: float64

<h1>Correlation Analysis</h1>
<hr>
<h2>Determining Correlation Using Pandas Pearson R</h2>

<h3>Is there a correlation between bill length and bill depth?</h3>

In [30]:
bill_spe_corr = penguin['bill_length_mm'].corr(penguin['bill_depth_mm'])

In [31]:
bill_spe_corr

np.float64(-0.23505287035553257)

<h4>Question 1: What species are included and how many penguins per species?</h4>

In [32]:
penguin['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

<h5>Insights: The most common species is Adelie with 152 penguins in the dataset.</h5>

<h4>Question 2: What is the average flipper length of each species?</h4>

In [33]:
penguin.groupby('species')['flipper_length_mm'].mean()

species
Adelie       190.025758
Chinstrap    195.823529
Gentoo       217.055768
Name: flipper_length_mm, dtype: float64

<h5>Insights: Gentoo penguins have the longest average flippers, while Adelie have the shortest.</h5>

<h4>Question 3: What is the median body mass for each species?</h4>

In [34]:
penguin.groupby('species')['body_mass_g'].median()

species
Adelie       3700.0
Chinstrap    3700.0
Gentoo       5000.0
Name: body_mass_g, dtype: float64

<h5>Insights: Gentoo penguins are the heaviest, while Adelie and Chinstrap are lighter.</h5>

<h4>Question 4: What is the average bill length by island?</h4>

In [35]:
penguin.groupby('island')['bill_length_mm'].mean()

island
Biscoe       45.249535
Dream        44.167742
Torgersen    39.046576
Name: bill_length_mm, dtype: float64

<h5>Insights: Penguins on Biscoe Island have the longest bills on average.</h5>

<h4>Question 5: What is the correlation between body mass and flipper length?</h4>

In [36]:
mass_flipp_corr = penguin['body_mass_g'].corr(penguin['flipper_length_mm'])

In [37]:
mass_flipp_corr

np.float64(0.8712017673060113)

<h5>Insight: There is a strong positive correlation: heavier penguins tend to have longer flippers.</h5>

<h4>Question 6: What is the average body mass grouped by sex?</h4>

In [38]:
penguin.groupby('sex')['body_mass_g'].mean()

sex
female    3862.272727
male      4518.220339
Name: body_mass_g, dtype: float64

<h5>Insight: Male penguins are generally heavier than females.</h5>

<h4>Question 7: What is the distribution of penguins over the years?</h4>

In [39]:
penguin['year'].value_counts().sort_index()

year
2007    110
2008    114
2009    120
Name: count, dtype: int64

<h5>Insight: The dataset is balanced across 2007–2009, with similar sampling each year.</h5>

<h4>Question 8: Which species has the widest bills on average?</h4>

In [40]:
penguin.groupby('species')['bill_depth_mm'].mean()

species
Adelie       18.346358
Chinstrap    18.420588
Gentoo       14.982114
Name: bill_depth_mm, dtype: float64

<h5>Insight: Chinstrap penguins have the widest bills, while Gentoo have the narrowest.</h5>

<h4>Question 9: Which species has the highest average body mass by island?</h4>

In [41]:
penguin.groupby(['island','species'])['body_mass_g'].mean()

island     species  
Biscoe     Adelie       3709.659091
           Chinstrap            NaN
           Gentoo       5076.016260
Dream      Adelie       3688.392857
           Chinstrap    3733.088235
           Gentoo               NaN
Torgersen  Adelie       3706.372549
           Chinstrap            NaN
           Gentoo               NaN
Name: body_mass_g, dtype: float64

<h5>Insights: Gentoo penguins on Biscoe Island are the heaviest overall.</h5>

<h4>Question 10: What is the average bill length per species and sex?</h4>

In [42]:
penguin.groupby(['species','sex'])['bill_length_mm'].mean()

species    sex   
Adelie     female    37.257534
           male      40.273695
Chinstrap  female    46.573529
           male      51.094118
Gentoo     female    45.563793
           male      49.156393
Name: bill_length_mm, dtype: float64

<h5>Insights: Males generally have longer bills than females across all species.</h5>

<h4>Question 11: What is the average body mass per species per year?</h4>

In [43]:
penguin.groupby(['year','species'])['body_mass_g'].mean()

year  species  
2007  Adelie       3696.428571
      Chinstrap    3694.230769
      Gentoo       5070.588235
2008  Adelie       3742.000000
      Chinstrap    3800.000000
      Gentoo       5019.565217
2009  Adelie       3664.903846
      Chinstrap    3725.000000
      Gentoo       5140.697674
Name: body_mass_g, dtype: float64

<h5>Insights: Body mass is consistent over years, no large changes observed.</h5>

<h4>Question 12: Which penguins have flipper length between 200 mm and 210 mm?</h4>

In [44]:
penguin[(penguin['flipper_length_mm'] >=200) & (penguin['flipper_length_mm'] <=210)]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
3       Adelie  Torgersen        43.92193            NaN         200.915205   
53      Adelie     Biscoe        42.00000           19.5         200.000000   
90      Adelie      Dream        35.70000           18.0         202.000000   
91      Adelie      Dream        41.10000           18.1         205.000000   
95      Adelie      Dream        40.80000           18.9         208.000000   
101     Adelie     Biscoe        41.00000           20.0         203.000000   
123     Adelie  Torgersen        41.40000           18.5         202.000000   
129     Adelie  Torgersen        44.10000           18.0         210.000000   
137     Adelie      Dream        40.20000           20.1         200.000000   
151     Adelie      Dream        41.50000           18.5         201.000000   
154     Gentoo     Biscoe        48.70000           14.1         210.000000   
157     Gentoo     Biscoe        46.50000           13.5         210.000000   
160     Gentoo     Biscoe        43.30000           13.4         209.000000   
166     Gentoo     Biscoe        45.80000           14.6         210.000000   
168     Gentoo     Biscoe        42.00000           13.5         210.000000   
170     Gentoo     Biscoe        46.20000           14.5         209.000000   
180     Gentoo     Biscoe        48.20000           14.3         210.000000   
183     Gentoo     Biscoe        42.80000           14.2         209.000000   
184     Gentoo     Biscoe        45.10000           14.5         207.000000   
190     Gentoo     Biscoe        44.00000           13.6         208.000000   
191     Gentoo     Biscoe        48.70000           15.7         208.000000   
192     Gentoo     Biscoe        42.70000           13.7         208.000000   
194     Gentoo     Biscoe        45.30000           13.7         210.000000   
198     Gentoo     Biscoe        45.50000           13.9         210.000000   
202     Gentoo     Biscoe        46.60000           14.2         210.000000   
204     Gentoo     Biscoe        45.10000           14.4         210.000000   
208     Gentoo     Biscoe        43.80000           13.9         208.000000   
210     Gentoo     Biscoe        43.20000           14.5         208.000000   
212     Gentoo     Biscoe        45.30000           13.8         208.000000   
228     Gentoo     Biscoe        47.50000           14.2         209.000000   
250     Gentoo     Biscoe        48.40000           14.4         203.000000   
258     Gentoo     Biscoe        41.70000           14.7         210.000000   
260     Gentoo     Biscoe        43.30000           14.0         208.000000   
261     Gentoo     Biscoe        48.10000           15.1         209.000000   
271     Gentoo     Biscoe        43.92193            NaN         200.915205   
289  Chinstrap      Dream        52.00000           18.1         201.000000   
291  Chinstrap      Dream        50.50000           19.6         201.000000   
302  Chinstrap      Dream        50.50000           18.4         200.000000   
303  Chinstrap      Dream        49.50000           19.0         200.000000   
305  Chinstrap      Dream        52.80000           20.0         205.000000   
307  Chinstrap      Dream        54.20000           20.8         201.000000   
309  Chinstrap      Dream        51.00000           18.8         203.000000   
313  Chinstrap      Dream        52.00000           20.7         210.000000   
315  Chinstrap      Dream        53.50000           19.9         205.000000   
316  Chinstrap      Dream        49.00000           19.5         210.000000   
321  Chinstrap      Dream        50.80000           18.5         201.000000   
327  Chinstrap      Dream        51.40000           19.0         201.000000   
329  Chinstrap      Dream        50.70000           19.7         203.000000   
333  Chinstrap      Dream        49.30000           19.9         203.000000   
334  Chinstrap      Dream        50.20000           18.8 

<h5>Insights: This shows that 50+ penguins have flipper length between 200 to 210mm</h5>

<h4>Question 13: Is there a correlation between bill length and bill depth?</h4>

In [45]:
bill_spe_corr = penguin['bill_length_mm'].corr(penguin['bill_depth_mm'])

In [46]:
bill_spe_corr

np.float64(-0.23505287035553257)

<h5>Insights: There is a negative correlation between bill length and bill depth</h5>